core/src/main/scala/org/apache/spark/resource/ResourceProfile.scala

```diff
@@ -0,0 +1,186 @@
+ * regular Spark configs with the prefix removed. For instance overhead memory in this api
+ * is memoryOverhead, which is spark.executor.memoryOverhead with spark.executor removed.
+ * Resources like GPUs are resource.gpu (spark configs spark.executor.resource.gpu.*)
+ *
+ * Executor:
+ *   memory - heap
+ *   memoryOverhead
+ *   pyspark.memory
+ *   cores
+ *   resource.[resourceName] - GPU, FPGA, etc
+ *
+ * Task requirements:
+ *   cpus
+ *   resource.[resourceName] - GPU, FPGA, etc
+ *
+ * This class is private now for initial development, once we have the feature in place
+ * this will become public.
+ */
+@Evolving
+private[spark] class ResourceProfile() extends Serializable {
+
+  private val _id = ResourceProfile.getNextProfileId
+  private val _taskResources = new mutable.HashMap[String, TaskResourceRequest]()
+  private val _executorResources = new mutable.HashMap[String, ExecutorResourceRequest]()
+
+  private val allowedExecutorResources = HashSet[String](
+    ResourceProfile.MEMORY,
+    ResourceProfile.OVERHEAD_MEM,
+    ResourceProfile.PYSPARK_MEM,
+    ResourceProfile.CORES)
+
+  private val allowedTaskResources = HashSet[String](ResourceProfile.CPUS)
+
+  def id: Int = _id
+
+  def taskResources: Map[String, TaskResourceRequest] = _taskResources.toMap
+
+  def executorResources: Map[String, ExecutorResourceRequest] = _executorResources.toMap
+
+  /**
+   * (Java-specific) gets a Java Map of resources to TaskResourceRequest
+   */
+  def taskResourcesJMap: JMap[String, TaskResourceRequest] = _taskResources.asJava
+
+  /**
+   * (Java-specific) gets a Java Map of resources to ExecutorResourceRequest
+   */
+  def executorResourcesJMap: JMap[String, ExecutorResourceRequest] = _executorResources.asJava
+
+
+  def reset(): Unit = {
+    _taskResources.clear()
+    _executorResources.clear()
+  }
+
+  def require(request: TaskResourceRequest): this.type = {
+    val rName = request.resourceName
+    if (allowedTaskResources.contains(rName) || rName.startsWith(RESOURCE_DOT)) {
+      _taskResources(request.resourceName) = request
+    } else {
+      throw new IllegalArgumentException(s"Task resource not allowed: ${request.resourceName}")
+    }
+    this
+  }
+
+  def require(request: ExecutorResourceRequest): this.type = {
+    val rName = request.resourceName
+    if (allowedExecutorResources.contains(rName) || rName.startsWith(RESOURCE_DOT)) {
+      _executorResources(request.resourceName) = request
+    } else {
+      throw new IllegalArgumentException(s"Executor resource not allowed: ${request.resourceName}")
+    }
+    this
+  }
+
+  override def toString(): String = {
+    s"Profile: id = ${_id}, executor resources: ${_executorResources}, " +
+      s"task resources: ${_taskResources}"
+  }
+}
+
+private[spark] object ResourceProfile extends Logging {
+  val UNKNOWN_RESOURCE_PROFILE_ID = -1
+  val DEFAULT_RESOURCE_PROFILE_ID = 0
+
+  val CPUS = "cpus"
+  val CORES = "cores"
```

Comment:
cpus and cores are really the same thing, right?  its unfortunate we already chose different names for task vs. executor in existing confs.

Reply:
sorry that was more just a general complaint from me about what we've already done, not really related to your change.  I think keeping it the same as the confs makes sense.